src/PollingBlockTracker.test.ts

```diff
@@ -3135,4 +3136,127 @@ describe('PollingBlockTracker', () => {
       );
     });
   });
+
+  describe('_newPotentialLatest', () => {
```

Comment:
Not sure where to stick these tests. Should they also exist on the SubscribeBlockTracker? Should they be moved into a new BaseBlockTracker spec?

Reply:
Do we actually have more than a polling block tracker? If not, this seems fine.

Reply:
As for where to put these tests, I feel that grouping them under `_newPotentialLatest` is a bit misleading, as this is a protected method, so the new behavior can't be exercised by calling that method; instead, it's activated by using any of the following strategies for getting blocks:

- `getLatestBlock`
- `checkForLatestBlock`
- listening to "latest" (can be done a few ways: `on`, `addListener`, `once`)
- listening to "sync" (can be done a few ways: `on`, `addListener`, `once`)

Another thing is that by separating the tests in this way, you're treating the new behavior as a layer that sits "on top" of the existing behavior, but this behavior is now an inherent part of all of the ways that you can get blocks and so it should be tested just like any other behavior.

Therefore, I wonder if it would make more sense to go into the relevant `describe` blocks for the strategies above and add the new tests you've added there. A couple things:

- These tests are as isolated as possible, so in the context of these `describe` blocks, the only thing being tested should be the strategy itself.
- Each strategy acts a little differently, so you may have to adapt the tests to fit the strategy.

As an example, say you add the tests to the `describe` block that represents `getLatestBlock`. This method is a bit unique in that it caches the latest block to avoid unnecessary requests. That means if you want to call it multiple times, you need to bust the cache before calling it again.

If you look at the existing tests for `getLatestBlock`, you should see an existing test "should ask for a new block number after the cached one is cleared". Since we are about to have multiple tests that bust the cache, it might make sense to put this in a `describe` that says so. Then we can add the new tests after that.

All said, this is what that could look like:

``` typescript
    describe('after the current block number cache is cleared', () => {
      it('should ask for a new block number when called again', async () => {
        const setTimeoutRecorder = recordCallsToSetTimeout();
        const blockTrackerOptions = {
          pollingInterval: 100,
          blockResetDuration: 200,
        };

        await withPollingBlockTracker(
          {
            provider: {
              stubs: [
                {
                  methodName: 'eth_blockNumber',
                  response: {
                    result: '0x0',
                  },
                },
                {
                  methodName: 'eth_blockNumber',
                  response: {
                    result: '0x1',
                  },
                },
              ],
            },
            blockTracker: blockTrackerOptions,
          },
          async ({ provider, blockTracker }) => {
            const sendAsyncSpy = jest.spyOn(provider, 'sendAsync');
            const blockNumber1 = await blockTracker.getLatestBlock();
            expect(blockNumber1).toStrictEqual('0x1');
            // When the block tracker stops, there may be two `setTimeout`s in
            // play: one to go to the next iteration of the block tracker
            // loop, another to expire the current block number cache. We don't
            // know which one has been added first, so we have to find it.
            await setTimeoutRecorder.nextMatchingDuration(
              blockTrackerOptions.blockResetDuration,
            );
            const blockNumber2 = await blockTracker.getLatestBlock();
            expect(blockNumber2).toStrictEqual('0x2');
            const requestsForLatestBlock = sendAsyncSpy.mock.calls.filter(
              (args) => {
                return args[0].method === 'eth_blockNumber';
              },
            );
            expect(requestsForLatestBlock).toHaveLength(2);
          },
        );
      });
    });

    it('should ignore a new block number that is less than the previous block number by default', async () => {
      recordCallsToSetTimeout();

      await withPollingBlockTracker(
        {
          provider: {
            stubs: [
              {
                methodName: 'eth_blockNumber',
                response: {
                  result: '0x1',
                },
              },
              {
                methodName: 'eth_blockNumber',
                response: {
                  result: '0x0',
                },
              },
            ],
          },
        },
        async ({ provider, blockTracker }) => {
          const sendAsyncSpy = jest.spyOn(provider, 'sendAsync');
          const blockNumber1 = await blockTracker.getLatestBlock();
          expect(blockNumber1).toStrictEqual('0x1');
          // When the block tracker stops, there may be two `setTimeout`s in
          // play: one to go to the next iteration of the block tracker
          // loop, another to expire the current block number cache. We don't
          // know which one has been added first, so we have to find it.
          await setTimeoutRecorder.nextMatchingDuration(
            blockTrackerOptions.blockResetDuration,
          );
          const blockNumber2 = await blockTracker.getLatestBlock();
          expect(blockNumber2).toStrictEqual('0x1');
          const requestsForLatestBlock = sendAsyncSpy.mock.calls.filter(
            (args) => {
              return args[0].method === 'eth_blockNumber';
            },
          );
          expect(requestsForLatestBlock).toHaveLength(2);
        },
      );
    });

    it('should not ignore a new block number that is less than the previous block number by default if the block tracker was initialized with `usePastBlocks: true', async () => {
      recordCallsToSetTimeout();

      await withPollingBlockTracker(
        {
          provider: {
            stubs: [
              {
                methodName: 'eth_blockNumber',
                response: {
                  result: '0x1',
                },
              },
              {
                methodName: 'eth_blockNumber',
                response: {
                  result: '0x0',
                },
              },
            ],
          },
          blockTracker: {
            usePastBlocks: true,
          },
        },
        async ({ provider, blockTracker }) => {
          const sendAsyncSpy = jest.spyOn(provider, 'sendAsync');
          const blockNumber1 = await blockTracker.getLatestBlock();
          expect(blockNumber1).toStrictEqual('0x1');
          // When the block tracker stops, there may be two `setTimeout`s in
          // play: one to go to the next iteration of the block tracker
          // loop, another to expire the current block number cache. We don't
          // know which one has been added first, so we have to find it.
          await setTimeoutRecorder.nextMatchingDuration(
            blockTrackerOptions.blockResetDuration,
          );
          const blockNumber2 = await blockTracker.getLatestBlock();
          expect(blockNumber2).toStrictEqual('0x0');
          const requestsForLatestBlock = sendAsyncSpy.mock.calls.filter(
            (args) => {
              return args[0].method === 'eth_blockNumber';
            },
          );
          expect(requestsForLatestBlock).toHaveLength(2);
        },
      );
    });
```

In this way you'd apply the same sort of strategy to the other `describe`s as well.

Let me know if any of that doesn't make sense and I'm happy to explain further.